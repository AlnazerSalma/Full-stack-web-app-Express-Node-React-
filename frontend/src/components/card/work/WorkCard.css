.workcard-container {
    display: flex;
    flex-direction: column;
}

.workcard-row {
    display: flex;
    justify-content: flex-start;
    align-items: center;
    padding: 12px;
}

.workcard-title {
    flex: 1;
    font-size: 2rem;
    font-weight: bold;
    padding-right: 12px;
    text-align: left;
}


.workcard-desc {
    flex: 1;
    font-size: 1.2rem;
    color: #ffffff;
    padding: 0 16px;
    text-align: left;
}

.workcard-year {
    flex: 0.5;
    font-size: 1.2rem;
    color: #ffffff;
    text-align: left;
}

.workcard-row:hover .workcard-title {
    color: #61dafb;
    transform: scale(1.05);
    transition: all 0.3s ease;
}

.workcard-media {
    width: 75%;
    height: 150px;
    object-fit: cover;
    border-radius: 8px;
}

.workcard-image-wrapper {
    flex: 1;
    display: flex;
    justify-content: flex-start;
    visibility: hidden;
    transform: translateY(0);
    transition: transform 0.3s ease, visibility 0.3s ease;
}

.workcard-row:hover .workcard-title+.workcard-image-wrapper,
.workcard-title:hover+.workcard-image-wrapper {
    visibility: visible;
    transform: translateY(-10px) scale(1.2);
}

/* Responsive adjustments */
@media (max-width: 768px) {
    .workcard-row {
        flex-direction: column;
        align-items: flex-start;
    }

    .workcard-title {
        width: 100%;
        font-size: 1.2rem;
        padding: 8px 0;
        font-weight: bold;
    }

    .workcard-subrow {
        display: flex;
        flex-direction: row;
        width: 70%;
    }

    .workcard-desc,
    .workcard-year {
        flex: 1;
        font-size: 0.8rem;
        padding: 8px 0;
    }

    .workcard-image-wrapper {
        justify-content: center;
        width: 70%;
        margin-top: 12px;
        padding-top: 7px;
        padding-bottom: 7px;
    }

    .workcard-media {
        width: 70%;
        height: 150px;
        padding-top: 7px;
        padding-bottom: 7px;
    }
}

@media (max-width: 480px) {
    .workcard-title,
    .workcard-desc,
    .workcard-year {
        font-size: 0.9rem;
    }

    .workcard-media {
        height: 100px;
    }
}
