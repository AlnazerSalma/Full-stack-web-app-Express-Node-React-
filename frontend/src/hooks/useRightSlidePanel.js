import { useState, useCallback } from "react";
function useRightSlidePanel() {
  const [selectedItem, setSelectedItem] = useState(null);
  const [isOpen, setIsOpen] = useState(false);

  const openPanel = useCallback((item) => {
    setSelectedItem(item);
    setIsOpen(true);
  }, []);

  const closePanel = useCallback(() => {
    setIsOpen(false);
    setSelectedItem(null);
  }, []);

  return {
    isOpen,
    selectedItem,
    openPanel,
    closePanel,
  };
}

export default useRightSlidePanel;
