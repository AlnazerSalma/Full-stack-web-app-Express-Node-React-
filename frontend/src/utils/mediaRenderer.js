import React from "react";

const mediaRenderer = ({ src, type, alt = "", className = "" }) => {
  if (!src) return null;

  if (type === "video" || src.endsWith(".mp4")) {
    return (
      <video
        src={src}
        autoPlay
        loop
        muted
        playsInline
        className={className}
      />
    );
  }

  return <img src={src} alt={alt} className={className} />;
};

export default mediaRenderer;
