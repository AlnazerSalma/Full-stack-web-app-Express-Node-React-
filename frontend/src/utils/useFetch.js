import { useEffect, useState } from "react";
import axios from "axios";

const useFetch = (url, maxRetries = 3, retryDelay = 1000) => {
  const [data, setData] = useState([]);
  const [loading, setLoading] = useState("Loading...");
  const [error, setError] = useState(null);

  useEffect(() => {
    let retries = 0;
    let isCancelled = false;

    const fetchData = async () => {
      setLoading("Loading...");
      while (retries < maxRetries) {
        try {
          const response = await axios.get(url);
          if (!isCancelled) {
            setData(response.data);
            setError(null);
            setLoading(null);
          }
          return;
        } catch (err) {
          retries += 1;
          if (retries >= maxRetries && !isCancelled) {
            setError("Failed to fetch data. Please try again later.");
            setLoading(null);
          } else {
            await new Promise((res) => setTimeout(res, retryDelay));
          }
        }
      }
    };

    fetchData();

    return () => {
      isCancelled = true;
    };
  }, [url, maxRetries, retryDelay]);

  return { data, loading, error };
};

export default useFetch;
