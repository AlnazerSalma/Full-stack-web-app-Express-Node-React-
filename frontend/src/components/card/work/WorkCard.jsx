import React, { useState} from "react";
import './WorkCard.css';
import WorksSliderPanel from "../../right_slider_panel/WorksCareerRightSlidePanel";
import useFetchData from "../../../utils/useFetchData";


function WorkCard() {
  const { data: works, loading, error } = useFetchData("http://localhost:5000/api/works");
  const [selectedItem, setSelectedItem] = useState(null);
  const [isPanelOpen, setIsPanelOpen] = useState(false);

  const handleRowClick = (item) => {
    setSelectedItem(item);
    setIsPanelOpen(true);
  };

  const handleClosePanel = () => {
    setIsPanelOpen(false);
    setSelectedItem(null);
  };
   if (loading) {

    return <p className="loading-message">Loading ...</p>

  }

  if (error) {

    return <p className="error-message">{error}</p>
  }

  return (
    <>
      <div className="workcard-container">
        {works.map((item, index) => {
          const firstMedia = item.media?.[0];

          return (
            <div className="workcard-row" key={index} onClick={() => handleRowClick(item)}>
              <div className="workcard-title">{item.title}</div>
              <div className="workcard-image-wrapper">
                {firstMedia?.type === "video" ? (
                  <video
                    src={firstMedia.src}
                    autoPlay
                    muted
                    loop
                    playsInline
                    className="workcard-media"
                  />
                ) : (
                  <img
                    src={firstMedia?.src || item.image}
                    alt={item.title}
                    className="workcard-media"
                  />
                )}
              </div>
              <div className="workcard-desc">{item.type}</div>
              <div className="workcard-year">{item.year}</div>
            </div>
          );
        })}
      </div>

      {selectedItem && (
        <WorksSliderPanel
          isOpen={isPanelOpen}
          onClose={handleClosePanel}
          role={selectedItem}
        />
      )}
    </>
  );
}

export default WorkCard;
