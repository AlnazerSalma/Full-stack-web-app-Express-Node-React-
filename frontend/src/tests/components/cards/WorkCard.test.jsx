import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import WorkCard from "../../../components/card/work/WorkCard";
import useFetch from "../../../utils/useFetch";
import useRightSlidePanel from "../../../hooks/useRightSlidePanel";
import mediaRenderer from "../../../utils/mediaRenderer";

jest.mock("../../../utils/useFetch");
jest.mock("../../../hooks/useRightSlidePanel");

jest.mock("../../../utils/mediaRenderer", () => ({
  __esModule: true,
  default: jest.fn(({ src, alt, className }) => (
    <img src={src} alt={alt} className={className} />
  )),
}));

jest.mock("../../../components/right_slider_panel/WorksRightSlidePanel", () => (props) => (
  <div data-testid="works-slider-panel">
    WorksSliderPanel is open: {props.isOpen ? "Yes" : "No"}
    <button onClick={props.onClose}>Close</button>
  </div>
));


describe("WorkCard Component", () => {
  const fakeWorks = [
    {
      title: "Work 1",
      media: [{ src: "media1.jpg", type: "image" }],
      image: "fallback1.jpg",
      type: "Type 1",
      year: "2020",
    },
    {
      title: "Work 2",
      media: [],
      image: "fallback2.jpg",
      type: "Type 2",
      year: "2021",
    },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("renders loading message when loading", () => {
    useFetch.mockReturnValue({ data: null, loading: "Loading works...", error: null });
    useRightSlidePanel.mockReturnValue({
      isOpen: false,
      selectedItem: null,
      openPanel: jest.fn(),
      closePanel: jest.fn(),
    });

    render(<WorkCard />);
    expect(screen.getByText("Loading works...")).toBeInTheDocument();
  });

  test("renders error message when error occurs", () => {
    useFetch.mockReturnValue({ data: null, loading: false, error: "Failed to fetch" });
    useRightSlidePanel.mockReturnValue({
      isOpen: false,
      selectedItem: null,
      openPanel: jest.fn(),
      closePanel: jest.fn(),
    });

    render(<WorkCard />);
    expect(screen.getByText("Failed to fetch")).toBeInTheDocument();
  });

  test("renders work cards and responds to click", async () => {
    const openPanelMock = jest.fn();
    useFetch.mockReturnValue({ data: fakeWorks, loading: false, error: null });
    useRightSlidePanel.mockReturnValue({
      isOpen: true,
      selectedItem: fakeWorks[0],
      openPanel: openPanelMock,
      closePanel: jest.fn(),
    });

    render(<WorkCard />);

    // Check that works are rendered
    expect(screen.getByText("Work 1")).toBeInTheDocument();
    expect(screen.getByText("Work 2")).toBeInTheDocument();

    // Check that mediaRenderer was called (image rendered)
    expect(mediaRenderer).toHaveBeenCalled();

    // eslint-disable-next-line testing-library/no-node-access
    const firstWorkCard = screen.getByText("Work 1").closest(".workcard-row");
    expect(firstWorkCard).toBeInTheDocument();
    await userEvent.click(firstWorkCard);
    expect(openPanelMock).toHaveBeenCalledWith(fakeWorks[0]);
    expect(await screen.findByTestId("works-slider-panel")).toBeInTheDocument();
  });
  test("matches snapshot when works are loaded", () => {
  useFetch.mockReturnValue({ data: fakeWorks, loading: false, error: null });
  useRightSlidePanel.mockReturnValue({
    isOpen: false,
    selectedItem: null,
    openPanel: jest.fn(),
    closePanel: jest.fn(),
  });

  const { asFragment } = render(<WorkCard />);
  expect(asFragment()).toMatchSnapshot();
});

});
