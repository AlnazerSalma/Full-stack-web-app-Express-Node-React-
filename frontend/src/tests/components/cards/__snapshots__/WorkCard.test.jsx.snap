// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkCard Component matches snapshot when works are loaded 1`] = `
<DocumentFragment>
  <div
    class="workcard-container"
  >
    <div
      class="workcard-row"
    >
      <div
        class="workcard-title"
      >
        Work 1
      </div>
      <div
        class="workcard-image-wrapper"
      />
      <div
        class="workcard-desc"
      >
        Type 1
      </div>
      <div
        class="workcard-year"
      >
        2020
      </div>
    </div>
    <div
      class="workcard-row"
    >
      <div
        class="workcard-title"
      >
        Work 2
      </div>
      <div
        class="workcard-image-wrapper"
      />
      <div
        class="workcard-desc"
      >
        Type 2
      </div>
      <div
        class="workcard-year"
      >
        2021
      </div>
    </div>
  </div>
</DocumentFragment>
`;
